realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java

```diff
@@ -1411,6 +1415,101 @@ public void onChange(RealmResults<AllTypes> results) {
         });
     }
 
+    @Test
+    @UiThreadTest
+    public void executeTransaction_mainThreadWritesAllowed() {
+        RealmConfiguration configuration = configFactory.createConfigurationBuilder()
+                .allowWritesOnUiThread(true)
+                .build();
+
+        Realm realm = Realm.getInstance(configuration);
+        realm.executeTransaction(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.insert(new Dog("Snuffles"));
+            }
+        });
+
+        RealmResults<Dog> results = realm.where(Dog.class).equalTo("name", "Snuffles").findAll();
+        assertEquals(1, results.size());
+        assertNotNull(results.first());
+        assertEquals("Snuffles", Objects.requireNonNull(results.first()).getName());
+
+        realm.close();
+    }
+
+    @Test
+    @UiThreadTest
+    public void executeTransaction_throwsWhenRunningOnMainThread() {
+        RealmConfiguration configuration = configFactory.createConfigurationBuilder()
+                .build();
+
+        // Try-with-resources
+        try (Realm realm = Realm.getInstance(configuration)) {
+            realm.executeTransaction(new Realm.Transaction() {
+                @Override
+                public void execute(Realm realm) {
+                    // no-op
+                }
+            });
+            fail("It is not allowed to run blocking transactions on the UI thread by default.");
+        } catch (RealmException e) {
+            assertTrue(Objects.requireNonNull(e.getMessage()).contains("allowWritesOnUiThread"));
+        }
+    }
+
+    @Test
+    public void executeTransaction_runsOnNonUiThreadRegardlessOfSetting() {
+        RealmConfiguration configuration = configFactory.createConfigurationBuilder()
+                .build();
+
+        Realm realm = Realm.getInstance(configuration);
+        realm.executeTransaction(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                // no-op
+            }
+        });
+    }
+
+    @Test
+    @UiThreadTest
+    public void query_runOnMainThreadAllowed() {
+        RealmConfiguration configuration = configFactory.createConfigurationBuilder()
+                .build();
+
+        Realm realm = Realm.getInstance(configuration);
+        realm.where(Dog.class).findAll();
+        realm.close();
+    }
+
+    @Test
+    @UiThreadTest
+    public void query_runOnMainThreadThrows() {
```

Comment:
With this setting, you also need to check and throw if `Realm.refresh()` is being called, because that will convert async queries to sync queries which would violate this restriction.

Reply:
Yes, we discussed that yesterday. I'm still working on it 👍